modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TestTcpDiscoverySpi.java

```diff
@@ -44,4 +52,19 @@
     @Override public void simulateNodeFailure() {
         super.simulateNodeFailure();
     }
+
+    /** {@inheritDoc} */
+    @Override public void setListener(@Nullable DiscoverySpiListener lsnr) {
+        super.setListener(lsnr ==  null || discoveryHooks == null ? lsnr : wrap(lsnr, discoveryHooks));
+    }
+
+    /**
+     * Sets interceptors of discovery messages.
```

Comment:
I think we should add javadoc and assert that hook should be set before Discovery started.